Remove delay-file-viewer flag, remove async file extraction, improve file-viewer consistency

EnTeQuAk opened the pull request:
The frontend still supports it, I didn't want to touch untested code tbh.

Fixes #4203, refs (or fixes) #3579

* [x] Lock the file-viewer more vigilant to avoid multiple extractions
* [x] Don't silently fall-back file/directory extraction - could be too verbose but that's better than partially extracted XPIs
* [x] Remove async file extraction as it has been unused for years apparently
* [x] Actually show extraction errors in the file-viewer... (see screenshot below)
* [x] Tests tests tests...

![screenshot from 2017-01-05 19-46-45](https://cloud.githubusercontent.com/assets/139033/21693233/708d5cc8-d380-11e6-81ae-cec0df4a7afb.png)

EnTeQuAk commented on src/olympia/lib/settings_base.py:
```diff
@@ -1137,9 +1137,6 @@ def JINJA_CONFIG():
     'olympia.editors.tasks.send_mail': {'queue': 'editors'},
 
-    # Files
-    'olympia.files.tasks.extract_file': {'queue': 'files'},
```
cc @jasonthomas that he doesn't get any notifications once this lands that the `files` queue stays empty.

EnTeQuAk commented:
@diox if you're up for it, this is ready for another review. I think it's done now :)

mstriemer commented on src/olympia/files/helpers.py:
```diff
@@ -60,6 +64,32 @@ def file_tree(files, selected):
+            msg.save(info_msg)
+    except Exception, err:
+        if settings.DEBUG:
+            msg.save(_('There was an error accessing file %s. %s.') %
```
Might be better to append the error to the translated string rather than getting both translated.

EnTeQuAk replied:
The error won't be in the translated string but yeah, we could potentially append it. Not sure if that makes a difference, I usually like having all arguments around in the translation too

mstriemer replied:
It looks like this is creating two strings to be translated though. One with `%s.` at the end and another, that's only used with `settings.DEBUG`, that has `%s. %s.`.

So I think both strings would get extracted but we don't really need to translate the two param version since it's only for development.

EnTeQuAk replied:
Ahhh, great catch! Again! The joys of copying things and not noticing details like that. it'd be nice to switch between details and not based on permission but for now I'll change that string and just append the error.

mstriemer commented on src/olympia/files/helpers.py:
```diff
@@ -78,34 +108,73 @@ def __init__(self, file_obj):
+            yield True
+        else:
+            # Not yet locked, save flag and delete on exit.
+            msg.save(True, time=LOCKED_TIMEOUT)
```
Seems unlikely but I think there's a race condition here.

Is there a specific package that could be used to do this? If there was a `msg.set_if_matches(value=True, match=False)` and it was supported by this store (elasticsearch?) that'd do the trick.

Redis has a `GETSET` command that would do the trick. `was_locked = msg.getset('true')  # if was_locked == 'true' we didn't acquire the lock.`

EnTeQuAk replied:
it's using memcached behind the scenes so it's not possible to do this without the race condition. I'll document it though, might be worth extracting that in the future and re-implementing it as a general helper-utility that works around race-conditions.

mstriemer replied:
It looks like memcached has a `cas` command for "check and set", it will only set the value if it is the same as when you last read it. Guido writes about it [1] and it looks like you could replace the `set()` call with `cas()` and `get()` with `gets()` if our helpers support it and handle the "it was changed" case as it already being locked.

[1] http://neopythonic.blogspot.ca/2011/08/compare-and-set-in-memcache.html

EnTeQuAk replied:
I keep forgetting the CAS command. I'll look into it, thanks for pointing it out!

diox replied:
Django cache backends support `get_or_set`... in django 1.9...

EnTeQuAk replied:
Hmm, let's keep this maybe in mind then and introduce a proper helper once we're on Django 1.11 LTS later this year (hopefully)

mstriemer commented on src/olympia/files/helpers.py:
```diff
@@ -78,34 +108,73 @@ def __init__(self, file_obj):
+
+        if msg.get():
+            # Already locked
+            yield True
```
This value feels reversed to me. I would imagine `lock()` would return `True` if you attained the lock not if it was already locked.

Would be good to at least comment about the return value in the docstring if you'd prefer to keep it this way.

EnTeQuAk replied:
True, I think I'll change that. I did have some reasons but thinking about it, using `False` as a return value for "wasn't locked but now it is" doesn't quite feel right.

mstriemer commented on src/olympia/files/helpers.py:
```diff
@@ -78,34 +108,73 @@ def __init__(self, file_obj):
+        If locked, no one is allowed to run any extraction in `self.src` to
+        avoid files / directories get messed up.
+        """
+        cache_key = self._cache_key(LOCKED)
```
This could use `self.is_locked()` in its current state. Actually looks like `is_locked()` is never used so it can be removed if you'd prefer.

EnTeQuAk replied:
great catch! I'll change this to use self.is_locked() now that I added it. I feel like it's a nice addition to the API.

mstriemer commented on src/olympia/files/views.py:
```diff
@@ -98,8 +97,7 @@ def browse(request, viewer, key=None, type='file'):
-    if (not waffle.switch_is_active('delay-file-viewer') and
-            not viewer.is_extracted()):
+    if not viewer.is_extracted():
         extract_file(viewer)
```
I'm not sure I understand why it isn't delaying, but if this is what we want I guess that's cool. Why remove this from celery?

EnTeQuAk replied:
The thing is, the 'delay-file-viewer' flag isn't set on prod or anywhere and possibly never was. So we're not using async extraction at the moment (unless I'm missing something) and I found, if it's not used, let's remove it until we might rethink the way we do it and then do it right

EnTeQuAk commented:
> I don't understand how doing this synchronously will prevent any errors. Wouldn't the locking fix the problems if this is left asynchronous?

I removed asynchronous because it was unused (because of the flag not being enabled on prod). The locking would decrease any potential risk but it's safer to just use the code that is running in prod.

Is this a slow operation? The error message suggests it could take 5 minutes so that doesn't seem like something we'd want to tie a web worker up for a long time doing (could this DDoS us?). Maybe I don't understand what's going on.

I added some comments inline but don't feel super confident with an r+ or r- so I'll just leave this here.

EnTeQuAk commented on static/js/zamboni/files.js:
```diff
@@ -1032,14 +1032,19 @@ $(document).ready(function() {
+                        $('#validating').after($('<div>', {
+                            'class': 'notification-box error',
+                            'text': json.msg[k]
+                        }));
```
Small note on this: The box isn't being shown / created if there are errors so we have to create the notification-box ourselves.

diox replied:
This should be a comment :)

EnTeQuAk commented:
Alrighty, addressed all review comments.

Generally, it'd be nice to land this tomorrow CET if possible to maybe deploy it later on Friday, if we don't feel like it it appears that it will be fine to deploy early next week.

The general issue here is that the file-viewer is currently not showing all errors / exceptions and that there can be files / directories only partially extracted which isn't obvious to the reviewer and a potential security issue since it's not easy to spot.

diox commented:
I agree with both of you FWIW: ideally it should be async, but this is a step in the right direction, improving locking and general behavior while keeping things synchronous like it's on prod right now.

EnTeQuAk commented:
We found the problem with messages getting pushed to the queue: It's actual evidence of the file system errors we were seeing earlier, creating a weird state where viewer.is_extracted() returns False (or just a race condition) and thus pushes the file-extraction task into the queue. This happened twice today, so it's relatively rare but often enough to be kinda scared.

These errors now get pushed to the end-user, they should get reduced but at least they don't end up creating unreasonable results but clearly communicate if something went wrong.
